### DISCUSSION TITLE
Modify browser location from server

### DISCUSSION THREAD

**Discussion by mwx23** [lona-web-org/lona#158] [Q&A] [answered]
This is a feature request.

My use-case is that I have a menu to select items from a route `/items/<item_id>`.

I can of course do with `<a href='/items/123' />`. This however results in a page reload.

I am currently using `<button/>`s with a server-side function that just replaces item in the DOM, mostly because I can with `lona` :)

The only issue with that is it would be nice if the server-side could update the location in the browser if the user wishes to bookmark that page.

Basically the ability to send data to the browser `history.replaceState()` API.

**Comment by mwx23** [comment]
Sorry I just realised this should be over in the Feature Request section of Discussions.

**Reply by fscherf** [reply]
No worries, converted it :)

**Comment by fscherf** [comment]
Hm i am not sure if i want such a feature. Let's say there was ``LonaView.set_url()``. Should you able to set the URL to ``/`` if the "real" URL is ``/foo/``? Should you be able to extend the URL within the current URL as a prefix like ``/foo/`` to ``/foo/bar/``? Also what should happen it the server has a route for ``/foo/bar/``? What should happen if the back button or the refresh button is clicked?

It adds a whole new dimension to debugging as well. If a user states "i have a problem with the view behind /foo/bar/" you can't just look in the routing table to find the broken view.

Seems like a lot added complexity to save a few redirects. Also in my experience it's a bad idea to tinker with the browser history for the most part. It gets confusing very fast, therefore i tried to emulate the behavior a siple flask app would have (and I needed 3 attempts to get that right)

If you really want that you ca, add that as a custom feature to your frontend template. I can give you a code example if you want

**Reply by mwx23** [reply]
You make very sane points :)

I realise I am abusing RESTfulness.

If I was to generalise this request:

- Can I send custom `lona` events?
- How would I hook a frontend script onto that custom event?

And now that I type that, is this possible right now with a frontend widget? 

In defense of the specific case of mutating the displayed URL, if I'm mutating the URL server-side yes I would have to be responsible for merging any external changes from a request. 

But if I'm being honest the beauty of `lona` is that I *can* mutate everything from the server-side. 

I was toying with a middleware layer that intercepted routes and allowed the single main View to just modify the DOM instead of a delivering a whole new View. 

In the end I guess what I'm more going for is a kiosk-style app, but that's just my current obsession :)

Any guidance on how to extend `lona` in this fashion would be greatly appreciated.

**Reply by fscherf** [reply]
> In defense of the specific case of mutating the displayed URL, if I'm mutating the URL server-side yes I would have to be responsible for merging any external changes from a request.
> 
> But if I'm being honest the beauty of `lona` is that I _can_ mutate everything from the server-side.

I am not gonna argue with that :) If you know what you are doing, Lona should not restrict you!
> Any guidance on how to extend `lona` in this fashion would be greatly appreciated.

Sure thing! I am happy to help!

> * Can I send custom `lona` events?
> * How would I hook a frontend script onto that custom event?
> 
> And now that I type that, is this possible right now with a frontend widget?

What you are searching for are [Custom Messages](https://lona-web.org/end-user-documentation/frontends.html#sending-custom-messages).

Because you can't open multiple websockets to the same server, Lona uses messages prefixed ``lona:`` for all it's internal messages, to leave room for application specific websocket messages. To send a custom websocket message you use [LonaView.send_str()](https://lona-web.org/end-user-documentation/views.html#lonaview-send-str-string-broadcast-false-filter-connections-lambda-connection-true-wait-true), and to handle it in the frontend ``lona_context.add_message_handler()``.

Full example: [Sending messages from server to client](https://lona-web.org/end-user-documentation/frontends.html#server-to-client)

The example implements the javascript call ``alert()``. If you send a message, starting with ``custom-message:`` the frontend catches this message and runs ``alert()`` (as i reread the documentation i notice that it should be improved)

**Reply by maratori** [reply]
What about following use case?
There is a page `/stock/<ticker>` to show some info about stocks.
This page also contains a drop down to select different ticker.
Is it possible to do redirection inside `Select.handle_change()`?
What is the most Lona-way to implement that?

**Reply by fscherf** [reply]
@maratori I am sorry, that's currently not possible. Only ``LonaView.handle_request()`` and ``Middleware.handle_request()`` can send redirects, because redirects are implemented as responses. 

But maybe that's a good idea

**Reply by fscherf** [reply]
@maratori I actually have a use case for redirects from ``handle_change()`` or ``handle_click()`` myself. Should we open a feature-request for this?

**Reply by maratori** [reply]
@fscherf Yes, why not?

**Reply by fscherf** [reply]
@maratori it's actually not that hard to implement. I created #159 (it's not the nicest code, and i have to write documentation for the release, but for now it's ok i guess)

**Comment by fscherf** [comment]
@mwx23 I changed the label to Q&A, because it's not a feature request anymore :)

**Comment by mwx23** [comment, marked as answer]
Just an update on my implementation for those interested.

Also this solution means you should be able to run other DOM methods (completely untested, buyer beware...). 
Obviously more care needs to be taken with the deserialisation from python. Perhaps a more formalised JSON structure.

```javascript
// frontend.js
    lona_context.add_message_handler(function (lona_context, raw_message) {
        if (!raw_message.startsWith('custom-message:')) {
            return raw_message;
        }
        let [namespace, method, ...args] = raw_message.split(':');
        let cmd = `${method}(...args)`;
        let res = eval(cmd);
        // serialise res and send it back to the server 
        // lona_context.send(`custom-message:${serialisedRes}`);
    });
```

Then in my View:
```python
self.send_str(f'custom-message:window.history.replaceState::my new tab title:/my/updated/route')
```

Absolutely loving how easy `lona` is to extend!

**Reply by fscherf** [reply]
@mwx23 Looks good! I do something similar in some of my work related projects to implement desktop notifications.
I am very glad Lona seems not to be in your way when you try to solve problems :)
